File: TEK3/Java/Java_jcoinche_2017/modules/rules/src/main/java/com/jcoinche/rules/Card.java
package com.jcoinche.rules;

/**
 * This class represents a card by defining all of its characteristics.
 *
 * A card has a value and a color.
 * The value of a card goes from the Seven to the Ace.
 * The color of a card is either Spade, Heart, Club or Diamond.
 *
 * @author uberti_l
 * @author grella_c
 * @version 1.0
 */
public class Card {
    public enum CardName {
        Seven,
        Eight,
        Nine,
        Ten,
        Jack,
        Queen,
        King,
        Ace
    }
    public enum CardType {
        Spade,
        Heart,
        Club,
        Diamond
    }
    private int tabPowerAsset[] = {0, 1, 6, 4, 7, 2, 3, 5};
    private int tabPowerNonAsset[] = {0, 1, 2, 6, 3, 4, 5, 7};
    private int tabPointsAsset[] = {0, 0, 3, 4, 10, 11, 14, 20};
    private int tabPointsNonAsset[] = {0, 0, 0, 2, 3, 4, 10, 11};
    private int tabPointsAllAssets[] = {0, 0, 2, 3, 5, 7, 9, 14};
    private int tabPointsNoneAssets[] = {0, 0, 0, 2, 3, 4, 10, 19};

    public CardName name;
    public CardType type;

    public int      powerAsset;
    public int      powerNonAsset;

    public int      pointsAsset;
    public int      pointsNonAsset;
    public int      pointsAllAssets;
    public int      pointsNoneAssets;

    /**
     * Basic constructor which initialize the characteristics of the card.
     *
     * @param _name The card value
     * @param _type The card color
     * @param num Index of the card to determine its amount of points and its power
     */
    Card(CardName _name, CardType _type, int num) {
        name = _name;
        type = _type;
        powerAsset = tabPowerAsset[num];
        powerNonAsset = tabPowerNonAsset[num];
        pointsAsset = tabPointsAsset[num];
        pointsNonAsset = tabPointsNonAsset[num];
        pointsAllAssets = tabPointsAllAssets[num];
        pointsNoneAssets = tabPointsNoneAssets[num];
    }
}

File: TEK3/DOTNET/DOT_cardGames_2017/Server/GameRules/Card.cs
namespace Coinche.GameRules
{
    public enum CardColor
    {
        Heart,
        Spade,
        Diamond,
        Club
    }

    public enum CardValue
    {
        Seven = 0,
        Eight = 1,
        Nine = 2,
        Ten = 3,
        Jack = 4,
        Queen = 5,
        King = 6,
        Ace = 7
    }

    /// <summary>
    /// class defining a card
    /// </summary>
    public class Card
    {
        /// <summary>
        /// color of the card
        /// </summary>
        public CardColor Color { get; }

        /// <summary>
        /// value of the card
        /// </summary>
        public CardValue Value { get; }

        /// <summary>
        /// power of the card if its color is the asset one
        /// </summary>
        public int PowerAsset { get; }

        /// <summary>
        /// power of the card if it's not an asset
        /// </summary>
        public int PowerNonAsset { get; }

        /// <summary>
        /// points of the card if it's an asset
        /// </summary>
        public int PointsAsset { get; }

        /// <summary>
        /// points of the card if it's not an asset
        /// </summary>
        public int PointsNonAsset { get; }

        /// <summary>
        /// points of the card if all cards are assets
        /// </summary>
        public int PointsAllAssets { get; }

        /// <summary>
        /// points of the card if there is none assets
        /// </summary>
        public int PointsNoneAssets { get; }

        /// <summary>
        /// Basic constructor that initialize a card
        /// </summary>
        public Card(CardColor c, CardValue v)
        {
            Color = c;
            Value = v;

            int[] tabPowerAsset = { 0, 1, 6, 4, 7, 2, 3, 5 };
            int[] tabPowerNonAsset = { 0, 1, 2, 6, 3, 4, 5, 7 };
            int[] tabPointsAsset = { 0, 0, 3, 4, 10, 11, 14, 20 };
            int[] tabPointsNonAsset = { 0, 0, 0, 2, 3, 4, 10, 11 };
            int[] tabPointsAllAssets = { 0, 0, 2, 3, 5, 7, 9, 14 };
            int[] tabPointsNoneAssets = { 0, 0, 0, 2, 3, 4, 10, 19 };

            int num = (int)v;
            PowerAsset = tabPowerAsset[num];
            PowerNonAsset = tabPowerNonAsset[num];
            PointsAsset = tabPointsAsset[num];
            PointsNonAsset = tabPointsNonAsset[num];
            PointsAllAssets = tabPointsAllAssets[num];
            PointsNoneAssets = tabPointsNoneAssets[num];
        }
    }
}
